### PR TITLE
Change ParameterDefaultValue to use GetUninitializedObject instead of Activator.CreateInstance

### PR DESCRIPTION
**Pull request by eerhardt**
Activator.CreateInstance has the ability to call a parameterless constructor (if one is defined) on a value type. This is incorrect for ParameterDefaultValue. It is more correct to call GetUninitializedObject, which is the same as using `default(T)`.

### CODE REVIEW

**Comment by pakrym** [comment]
I think we should add a test.

File: src/libraries/Common/src/Extensions/ParameterDefaultValue/ParameterDefaultValue.cs [hunk 1]
```diff
@@ -39,21 +43,27 @@ public static bool TryGetDefaultValue(ParameterInfo parameter, out object? defau
                     defaultValue = parameter.DefaultValue;
                 }
 
+                bool isNullableParameterType = parameter.ParameterType.IsGenericType &&
+                    parameter.ParameterType.GetGenericTypeDefinition() == typeof(Nullable<>);
+
                 // Workaround for https://github.com/dotnet/runtime/issues/18599
-                if (defaultValue == null && parameter.ParameterType.IsValueType)
+                if (defaultValue == null && parameter.ParameterType.IsValueType
```
Comment (pakrym) [inline src/libraries/Common/src/Extensions/ParameterDefaultValue/ParameterDefaultValue.cs:50]: Why didn't we need this check before?

**Comment by pakrym** [comment]
LGTM

**Comment by eerhardt** [comment]
> I think we should add a test.

I'm not sure how to add a test without https://github.com/dotnet/csharplang/issues/99. I would have to write a value type in IL that defines a parameterless constructor. Maybe we could do it with `Microsoft.NET.Sdk.IL`, but I'm not sure how much work that would be. My plan was to wait for https://github.com/dotnet/csharplang/issues/99 to add a test. Thoughts?

File: src/libraries/Common/src/Extensions/ParameterDefaultValue/ParameterDefaultValue.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Common/src/Extensions/ParameterDefaultValue/ParameterDefaultValue.cs:50]: Previously we were calling `Activator.CreateInstance`.

```C#
Type t = typeof(Nullable<int>);
Console.WriteLine($"Activator Is Null? {Activator.CreateInstance(t) == null}");
Console.WriteLine($"GetUninitializedObject Is Null? {RuntimeHelpers.GetUninitializedObject(t) == null}");
```

prints

```
Activator Is Null? True
GetUninitializedObject Is Null? False
```

https://github.com/dotnet/runtime/pull/44843#issuecomment-729344082 says `System.Nullable<T> (will be treated as simply T)` for `GetUninitializedObject`.

**Comment by stephentoub** [comment]
We have a few IL test projects, e.g.
https://github.com/dotnet/runtime/tree/master/src/libraries/System.Runtime/tests/TestModule
https://github.com/dotnet/runtime/tree/master/src/libraries/System.Runtime/tests/TestStructs
https://github.com/dotnet/runtime/tree/master/src/libraries/System.Reflection.Metadata/tests/Resources/Interop
This one even has what you want :)
https://github.com/dotnet/runtime/blob/2ffb39e4779a8b4cfb3d17191f5e959d47ea41ea/src/libraries/System.Runtime/tests/TestStructs/System.TestStructs.il#L25-L37

**Comment by eerhardt** [comment]
Thanks @stephentoub! I didn't know about those tests. I added a test using that struct.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs [hunk 2]
```diff
@@ -7,6 +7,9 @@ namespace Microsoft.Extensions.DependencyInjection.Specification
 {
     public class UnityDependencyInjectionSpecificationTests: SkippableDependencyInjectionSpecificationTests
     {
+        // See https://github.com/unitycontainer/microsoft-dependency-injection/issues/87
+        public override bool ExpectStructWithPublicDefaultConstructorInvoked => true;
```
Comment (pakrym) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: You can use the `SkippedTests` for this.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: I didn't want to skip the whole test, I just wanted to change this one assert. It didn't seem important enough to add a whole new test. And, in my opinion, it feels more correct to verify the behavior, even if the behavior is not "correct".

Comment (pakrym) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: As long as 3rd party library authors understand that this is something they are expected to fix.

Comment (pakrym) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: We might want to cc all of them on this PR. @davidfowl had a list somewhere.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: Unity was the one with this behavior. I opened the above issue (which points back to this PR) to start the discussion on if this is the expected behavior in Unity or not.

**Comment by eerhardt** [comment]
Test failures are #47728

**Review by GrabYourPitchforks — COMMENTED** [review]
The logic in _ParameterDefaultValue.cs_ LGTM.

**Review by pakrym — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs (hunk 2 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Unity.cs:11]: based on https://github.com/dotnet/runtime/issues/35999 the alias for unity is @ENikS

**Review by maryamariyan — APPROVED** [review]

**Pull request merged**